src/components/epic/signup/logic/form.schema.ts

```diff
@@ -1,11 +1,19 @@
 import * as z from 'zod';
 
 export const schema = z.object({
+  userName: z.string().nonempty('This field is required'),
   firstName: z.string().nonempty('This field is required'),
   lastName: z.string().nonempty('This field is required'),
   password: z.string().nonempty('This field is required'),
   idRole: z.number().gte(0, 'You need to select a role'),
   idSkills: z.number().array().min(2, 'You need to select at least two skills'),
+}).refine(async (val) => {
+  await new Promise((r) => setTimeout(r, 1000));
+  console.log("validation result ", val.userName !== "potato");
+  return val.userName !== "potato";
+}, {
+  message: "username is already taken :(",
+  path: ["userName"]
 });
```

Comment:
Async Validation logic in Zod. A special "refine" method is appended to the schema that makes an API request to check if the username is available. for the POC the API request is replaced with a timeout, and the only user name that is already occupied is "potato".